streams/shared/src/main/scala/zio/stream/experimental/ZStream.scala

```diff
@@ -1388,15 +1388,31 @@ abstract class ZStream[-R, +E, +O](
       } yield pull
     }
 
+  /**
+   * Runs the sink on the stream to produce either the sink's result or an error.
+   */
+  def run[R1 <: R, E1 >: E, O1 >: O, B](sink: ZSink[R1, E1, O1, B]): ZIO[R1, E1, B] =
+    (process <*> sink.push).use {
+      case (pull, push) =>
+        def go: ZIO[R1, E1, B] = pull.foldCauseM(
+          Cause
+            .sequenceCauseOption(_)
+            .fold(
+              push(None).foldCauseM(
+                Cause.sequenceCauseEither(_).fold(IO.halt(_), IO.succeedNow),
+                _ => IO.dieMessage("empty stream / empty sinks")
```

Comment:
Couldn't find a way around this one, when the stream sends a stop signal and the sink does not produce a Z, it's clearly a bug.

Reply:
Yeah. That was the whole premise of SignalFunction, no?

Reply:
Yes. I think this will be the only place where this happens though.

Reply:
I think it's a reasonable price to pay to not have to carry around SignalFunction :-)

Won't it happen with all sinks though?

Reply:
Hmm, I'm not so sure. I wrote `collectAll` and `fromEffect` with two different styles: in the first, I keep track of the state of the machine and will never process elements once it's done, while in the latter I trust that the stream will tell me when things are over. I think the latter is better, it simplifies implementations by a lot. Still an implicit contract, but now all the responsibility is in `ZStream`.